matrix-client-images.el

```diff
@@ -0,0 +1,96 @@
+  "Download image from URL and insert it at message MESSAGE-ID in ROOM."
+  (request url
+           :parser (apply-partially #'matrix-client-parse-image room)
+           :success (apply-partially #'matrix-client-insert-image-callback
+                                     :room room
+                                     :message-id message-id
+                                     :url url)))
+
+(cl-defmethod matrix-client-parse-image ((room matrix-client-room))
+  "Parse image from current HTTP response buffer and return image object."
+  (pcase-let* ((data (progn
+                       ;; Disabling multibyte is required for reading binary data.
+                       ;; FIXME: require or autoload this function
+                       (mm-disable-multibyte)
+                       ;; Point is where the body starts, after the headers
+                       (buffer-substring (point) (point-max))))
+               ((eieio buffer) room))
+    (with-current-buffer buffer
+      ;; Rescale image in room buffer to get proper size
+      (matrix-client-rescale-image data))))
+
+(defun matrix-client-rescale-image (data)
+  "Rescale DATA, if too big, to fit the current buffer."
+  ;; Copied from image.el
+  (if (not (and (fboundp 'imagemagick-types)
+                (get-buffer-window (current-buffer))))
+      (create-image data nil t :ascent 100)
+    (let ((edges (window-inside-pixel-edges
+                  (get-buffer-window (current-buffer))))
+          (max-width (window-pixel-width))
+          (max-height (/ (window-pixel-height) 2)))
+      (create-image data 'imagemagick 'data-p
+                    :max-width max-width
+                    :max-height max-height))))
+
+(cl-defmethod matrix-client-insert-image-callback (&key (room matrix-client-room) message-id url
+                                                        data error-thrown symbol-status response
+                                                        &allow-other-keys)
+  "Insert image into proper place at URL in message MESSAGE-ID in ROOM.
+Image is passed from parser as DATA, which should be an image
+object made with `create-image'.  This function should be called
+as an async callback when the image is downloaded."
+  (with-slots (buffer) room
+    (with-current-buffer buffer
+      ;; Starting with last message, search backward to find message
+      (cl-loop initially do (goto-char (point-max))
+               for event_id = (get-text-property (point) 'event_id)
+               until (equal event_id message-id)
+               do (goto-char (previous-single-property-change (point) 'event_id)))
+      ;; Find beginning and end of message text
+      (let* ((beg (point))
+             (end (next-single-property-change beg 'event_id))
+             (inhibit-read-only t)
+             (props (text-properties-at beg))
+             (string (with-temp-buffer
+                       ;; NOTE: These space characters before the newlines
```

Comment:
Hmm, that's interesting, but I have no idea why that would happen :(

Reply:
Me neither.  It was really weird, and it didn't happen in the non-async branch.  I stumbled upon the solution by accident.  I suppose it might be worth reporting as an Emacs bug someday, but for now, this definitely works (on my config, anyway!).